### PR TITLE
Refactor RouteOptions.toUrl

### PR DESCRIPTION
**Pull request by korshaknn**
closes https://github.com/mapbox/mapbox-java/issues/1313

can't use`java.net.URL` case it doesn't have handy methods to build URL, so use okhttp `HttpUrl`

also some symbols were encoded:
; -> %3B
, -> %2C
it shouldn't be an issue, right?

### CODE REVIEW

File: services-directions-models/build.gradle [hunk 1]
```diff
@@ -2,6 +2,7 @@ apply plugin: 'java-library'
 
 dependencies {
     api project(":services-geojson")
+    api dependenciesList.okhttp3
```
[the remark this answers is not in the document]
Comment (korshaknn) [reply services-directions-models/build.gradle:5]: as I said  we can't use `java.net.URL` because it doesn't have handy methods to build URL, we will struggle with `/` symbols anyway, I haven't found any workaround so use okhttp `HttpUrl`

[the remark this answers is not in the document]
Comment (RingerJK) [reply services-directions-models/build.gradle:5]: let's migrate `toUrl` method to `service` module, smth like
``kotlin
class ServiceUtils {
  public static String toUrl(RouteOption routeOptions) { 
    ...
  }
}
```

and current fun. This way we don't need okhttp dependency in current module

[the remark this answers is not in the document]
Comment (RingerJK) [reply services-directions-models/build.gradle:5]: kotlin extensions would look better 💯

[the remark this answers is not in the document]
Comment (korshaknn) [reply services-directions-models/build.gradle:5]: @RingerJK 
we have 2 separate dependencies:
- "com.mapbox.mapboxsdk:mapbox-sdk-services:version"
- "com.mapbox.mapboxsdk:mapbox-sdk-directions-models:version" 

if we move `toUrl ` method to `services` module, customers will be forced to add `services` as a dependency  to use it, but it's a wrapper around a bunch of modules:
```
main.java.srcDirs = ['../services-directions/src/main/java', '../services-geocoding/src/main/java',
                         '../services-matching/src/main/java', '../services-matrix/src/main/java',
                         '../services-optimization/src/main/java', '../services-staticmap/src/main/java',
                         '../services-speech/src/main/java', '../services-tilequery/src/main/java',
                         '../services-route-tiles/src/main/java', '../services-directions-refresh/src/main/java',
                         '../services-isochrone/src/main/java']
```

so, I think it's better to have `okhttp` dependency in  `directions-models` module

cc @LukasPaczos @Guardiola31337 wdyt?

[the remark this answers is not in the document]
Comment (RingerJK) [reply services-directions-models/build.gradle:5]: sorry, I mean the module, that @Guardiola31337 mentioned above - `service-core` of course. I think it's fine. If you need to get **URL**, looks like you need to get some `url/http-depends` module

[the remark this answers is not in the document]
Comment (LukasPaczos) [reply services-directions-models/build.gradle:5]: The `directions-models` was extracted from the `services` module exactly to remove the dependency on `okhttp`. The models should have no dependencies on the networking code. My opinion is that we shouldn't depend on `okhttp` here.

[the remark this answers is not in the document]
Comment (LukasPaczos) [reply services-directions-models/build.gradle:5]: Are there any other alternatives?

[the remark this answers is not in the document]
Comment (RingerJK) [reply services-directions-models/build.gradle:5]: I see there 4 options: 
- add okhttp dependency to the `models` module;
- migrate `toUrl` method do the `services` module;
- support `toUrl` with `StringBuilder` (may be we can use `java.URI` + `StringBuilder` to restrict issues with url building);
- extend `toUrl` with a param like `toUrl(HttpBuilder)`. Provide as a separate library `HttpBuilder implementation` (it might event depends on different libraries or customer might implement it on his own).

[the remark this answers is not in the document]
Comment (korshaknn) [reply services-directions-models/build.gradle:5]: @LukasPaczos @RingerJK 
> Are there any other alternatives?

if we don't want to depend on `okhttp` code we have 2 options:
- move `toUrl` method to another module and make it static util fun.
- maintain existing `RouteOptions::toUrl` fun, there might be issues with symbols (like a crash with '//' after the base url), but tests should cover it


I can't say which option I like more
- as I said if we choose the first one, customers can't use this method without an extra dependency on `services` 
> I think it's fine. If you need to get URL, looks like you need to get some url/http-depends module

can't agree with it, it's useful to get url just to try the same request with a bowser/postman/etc. 
- the second one is harder to maintain, but we have added more tests

if I need to choose, I would take the second one and continue with `StringBuilder`.

[the remark this answers is not in the document]
Comment (RingerJK) [reply services-directions-models/build.gradle:5]: I vote to move `toUrl` to the `service` module. `Models` has to be as dumb as possible and assembling _URLs_ based on fields is not what models should make

[the remark this answers is not in the document]
Comment (korshaknn) [reply services-directions-models/build.gradle:5]: @LukasPaczos @Guardiola31337 vote please and I will continue with the approach

[the remark this answers is not in the document]
Comment (LukasPaczos) [reply services-directions-models/build.gradle:5]: We need to maintain the current version of `RouteOptions::toUrl` anyway, we don't want to release a new breaking `mapbox-java` release, so I would vote for this option. The method does not need to do any encoding, just a simple unencoded URL that HTTP clients can then interpret. We can migrate to string builders if that's easier and cover edge cases with tests.

**Comment by korshaknn** [comment]
I've fixed `//` issue and added a test.
cc @LukasPaczos @RingerJK

File: services-directions-models/build.gradle (hunk 1 above)
[the remark this answers is not in the document]
Comment (korshaknn) [reply services-directions-models/build.gradle]: ok, 2-1 wins `stringBuilder` + tests

**Review by RingerJK — APPROVED** [review]
LGTM

File: services-directions-models/build.gradle (hunk 1 above)
[the remark this answers is not in the document]
Comment (Guardiola31337) [reply services-directions-models/build.gradle]: Late to the party but I'd vote `StringBuilder` + tests too 😅
